src/main/java/com/linecorp/armeria/server/http/jetty/ArmeriaConnector.java

```diff
@@ -0,0 +1,406 @@
+        final MetaData.Response info = transport.info;
+        if (info == null) {
+            throw new IllegalStateException("response metadata unavailable");
+        }
+
+        final FullHttpResponse res = new DefaultFullHttpResponse(
+                io.netty.handler.codec.http.HttpVersion.HTTP_1_1,
+                HttpResponseStatus.valueOf(info.getStatus()),
+                content, false);
+
+        info.getFields().forEach(e -> res.headers().add(e.getName(), e.getValue()));
+
+        return res;
+    }
+
+    /**
+     * Adds the hostname string to the specified {@link InetSocketAddress} so that
+     * Jetty's {@code ServletRequest.getLocalName()} implementation returns the configured hostname.
+     */
+    InetSocketAddress addHostname(InetSocketAddress address) {
+        try {
+            return new InetSocketAddress(InetAddress.getByAddress(
+                    hostname, address.getAddress().getAddress()), address.getPort());
+        } catch (UnknownHostException e) {
+            throw new Error(e); // Should never happen
+        }
+    }
+
+    private static final class ArmeriaConnectionFactory implements ConnectionFactory {
+        @Override
+        public String getProtocol() {
+            return PROTOCOL_NAME;
+        }
+
+        @Override
+        public List<String> getProtocols() {
+            return PROTOCOL_NAMES;
+        }
+
+        @Override
+        public Connection newConnection(Connector connector, EndPoint endPoint) {
+            throw new UnsupportedOperationException();
+        }
+    }
+
+    private final class ArmeriaHttpChannel extends HttpChannel {
+        ArmeriaHttpChannel(ServiceInvocationContext ctx, HttpTransport transport) {
+            super(ArmeriaConnector.this, ArmeriaConnector.this.getHttpConfiguration(),
+                  new ArmeriaEndPoint(ctx.localAddress(), ctx.remoteAddress()), transport);
+        }
+    }
+
+    private static final class ArmeriaHttpTransport implements HttpTransport {
+
+        final ByteBuf out;
+        MetaData.Response info;
+        Throwable cause;
+
+        ArmeriaHttpTransport(ByteBuf out) {
+            this.out = out;
+        }
+
+        @Override
+        public void send(MetaData.Response info, boolean head,
+                         ByteBuffer content, boolean lastContent, Callback callback) {
+
+            if (info != null) {
+                this.info = info;
+            }
+
+            out.writeBytes(content);
+            callback.succeeded();
+        }
+
+        @Override
+        public boolean isPushSupported() {
+            return false;
+        }
+
+        @Override
+        public void push(MetaData.Request request) {}
+
+        @Override
+        public void onCompleted() {}
+
+        @Override
+        public void abort(Throwable failure) {
+            cause = failure;
+        }
+
+        @Override
+        public boolean isOptimizedForDirectBuffers() {
+            return false;
```

Comment:
Ideally, I would want to build a Netty composite buffer from the `ByteBuffer`s Jetty produced without copying anything, but I could not, because:

- It seems like Jetty creates a 32KiB buffer even if the content length is actually much smaller than that. It's huge amount of memory especially given that the receiving peer might not read fast enough. I had to create a copied buffer with smaller capacity.
- As soon as I return from `send()`, Jetty returns the `ByteBuffer` back to the pool, which means I need to make a copy to take it beyond `send()`.

So, assuming that there's no way to avoid a copy, copying a heap buffer into Netty's direct buffer shouldn't be a big deal because Netty internally uses Unsafe.copyMemory() which allows optimized memory copy between a byte[] and a direct buffer.